### DISCUSSION TITLE
How can I change a real  variable by comparing  sdpvar variables with real  variables?

### DISCUSSION THREAD

**Discussion by Abersa** [yalmip/YALMIP#1228] [Problems using YALMIP]
Hi~
There is  a array about price, just like:
``` m
Price.Tables(1,96) = [0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687];
```
Then, there is a sdpvar : eRE(1,96).
I want to turn Price.table(1,t)  into 0.3 (Price.Disc), when eRE(1,t)<0.
Its my code:
```
evElectionPrice = sdpvar(1,96,'full');     % Set the price as  sdpvar variable 
evElectionPriceMAX(1,1:96) = Price.Disc; % Price.Disc = 0.3. evElectionPriceMAX is maximum of price.
evElectionPriceMIN(1,1:96) = Price.Disc;  % evElectionPriceMIN is minimum of price.
% c is constraints
c = [c,evElectionPrice<=evElectionPriceMAX];  % I  want to change evElectionPriceMAX(1,t) and the evElectionPriceMIN(1,t) to fit eRE(1,t)
c = [c,evElectionPrice>=evElectionPriceMIN];
d = binvar(2,1);
for t = 1 : 96 % I cant  use the implies well
    d = binvar(4,1);
    c = [c,sum(d) == 1,
        implies(d(1),[eRE(1,t) > 0,evElectionPriceMAX(1,t) == Price.Table1(1,t)]);
        implies(d(2),[eRE(1,t) > 0,evElectionPriceMIN(1,t) == Price.Table1(21t)]);
        implies(d(3),[eRE(1,t) <= 0,evElectionPriceMIN(1,t) == Price.Disc]);
        implies(d(4),[eRE(1,t) <= 0,evElectionPriceMAX(1,t) == Price.Disc])];
end
objective = sum(evElectionPrice.*abs(eRE));
```
![屏幕截图 2023-03-09 170933](https://user-images.githubusercontent.com/127385900/223974852-e984f785-0010-4186-b3cb-53af07fc5d88.png)
Obviously, It cant run, but I cant revise code and dont know how to use implies.
Could u help?

**Comment by mrkento11** [comment]
I just read your code, and I don't know what it does, so it is very difficult to try to understand your reasoning. Moreover, don't know what the variables do.

However, my comments would be:
1. the variable d does not have the dimension of the amount of cases you have. it should be d = binvar(4,1). You also declared d = binvar() twice. Once outside the for loop and once inside.
2. Both the variable evElectionPriceMAX and evElectionPriceMIN, are the same (Price.Disc). This implies ElectionPrice is bounded up and down by the same number.
3. Your for loop seem to use d as a binvar, but then after every loop you overwrite this variable. I truly don't fully understand what you want to achieve, but it seems that you want to have four cases for 96 instances, so it seems that you need to declare d = binvar(4,96);
4. You have some warning because you use strict inequalities. Yalmip is gentle enough to try to correct your mistake, but they are not supported since a while ago. For example: {eRE(1,t) > 0 -> eRE(1,t) >= 0}. Refer to: https://yalmip.github.io/inside/strictinequalities/ 

Try to formulate the question again to give us better clues to help you out.

**Reply by Abersa** [reply]
Thank you for your reply.My explanation is not clear enough. It is my fault, so I will make a detailed explanation next.
There are 96 time periods in a day (15 minutes for each time period).When the power(**eRE**) is positive number in **t** time period, the electricity price of each period is the corresponding value in **Price. Table1**.When the power is non-integer, the price of corresponding **t** period becomes 0.3.
![屏幕截图 2023-03-10 084829](https://user-images.githubusercontent.com/127385900/224195350-9b03f080-c846-426a-b3c2-01d1968765af.png)
My objective function is ：
![屏幕截图 2023-03-10 093107](https://user-images.githubusercontent.com/127385900/224200262-c7425d10-6121-433a-9c71-bacd3b660239.png)
The power of each period is the variable I need to solve, so I set it to sdpvar.I used a stupid way to set the electricity price.Set the maximum and minimum value of electricity price in each period, and make adjustments according to the power. Make the maximum and minimum value of electricity price equal in each time period to ensure that the electricity price is constant.

**Reply by johanlofberg** [reply]
Your initial question is weird, as you talk about a constant that you want to change depending on the value of the decision variables. Then it isn't a constant but a function.

You want to use the term `c(x)*abs(x)` where `c(x) `is the piecewise constant function` c(x) = {a when x >0, b when x<=0}.` If the two constants a and b are non-negative, this is nothing but the convex picewise linear function` max(a*x,-b*x)` which you thus use when summing up the objective terms. If some constant is negative, the function is no longer convex, and you have model it using logic programming by introducing a new variable f to represent the cost term, two binary variables case1 and case2 to represent which sign you have, and then use the implications 

`implies(case1, [f == a*x, x>=0]) + implies(case2, f == b*x, x<= 0]) + [case1+case2 == 1]`

If you are lazy, you can simply use the term `a*max(x,0) + b*max(-x,0)` in the objective term, and the modelling layer will recognize the nonconvexity due to the negative sign and automatically derive the MILP-representation of the second problematic term. The explicit logic representation will most likely yield a slightly better model though.

https://yalmip.github.io/example/modellingif/

**Reply by mrkento11** [reply]
If I may add my two cents:
I would just divide the variable Power into EV_charging and EV_discharging. The function Price(t) is fixed, and the discharging price is fixed as well, so we just might have an objective function c1(t)*EV_charging + c2*EV_discharging.

```
clear; clc; yalmip('clear');

% Separate price 
Price_charging = [0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.365 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.869 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687 0.687];
Price_discharging = -0.3;

% General information
no_EV = 2;
time_periods = 96;
EV_battery_capacity = 100; % I just google EV battery capacity, but you might have a different modeling
% Decision Variables
EV_charging = sdpvar(no_EV,time_periods,'full');
EV_discharging = sdpvar(no_EV,time_periods,'full');

% Define Constraints
Constraints = Constraints +[0 <= EV_charging <= EV_battery_capacity];       % Bounds for the charging capacity 
Constraints = Constraints +[0 <= EV_discharging <= EV_battery_capacity];    % Bounds for the chargning capacity
% Most likely you will add some constraints about the relationship in the
% SOC of the batteries across time.
% These set of additional constraints in itself should make sure that the
% EV battery is not charging and discharging at the same time.

% However, you should still in fact check if the battery might be charging
% and discharging at the same time.

% Objective Function
Objective = 0;
for i = 1:no_EV
    Objective = Objective + Price_charging'*EV_charging(i,:);           % Add the cost of charging EVs
    Objective = Objective + Price_discharging'*EV_discharging(i,:);     % Add the price the grid would be paying to the EV by injecting energy
end

% Add some sdpsettings

optimize(Constraints,Objective,options);
```
This way, you actually might even bypass the need of implies() and be sure your problem is still a MILP by not adding nonlinear functions like max or abs.
Just would add the caveat that you need to pay attention to the additional constraints to make sure that the EV is not charging and discharging at the same time.

**Reply by johanlofberg** [reply]
Yes, that is basically the manual representation of the absolute value/max operator in `max(a*x,-b*x)`

**Reply by Abersa** [reply]
Thank u, I think I am trapped in a strange circle of thinking. I should regard electricity price as a function. Another question: Why does the calculation of the objective function become slow after using the absolute value？How to calculate variance in Yalmip？

**Reply by johanlofberg** [reply]
Why are you trapped. I gave you the model, and it is very basic as it is just a matter of a two implies

Not  sure what you mean with slow when adding abs. How abs enters the model can create vastly different models, depending on convexity (as YALMIP has to use a MILP representation if you use abs in a nonconvex way). However, I gave you a model which doesn't require any abs

**Reply by johanlofberg** [reply]
This though` evElectionPrice.*abs(eRE) `will be extremely bad, as both evElectionPrice and eRE are decision variables, so you end up with a bilinear expression (and an absolute value operator), and a much much harder problem

**Reply by Abersa** [reply]
I think I understand what you mean。Thanks
